File: Solution1.java
//Leetcode problem : https://leetcode.com/problems/search-a-2d-matrix/
// Time Complexity : O(log(mn))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
/*
A n * m matrix converted into an array: matrix[x][y] → a[x * m + y]
An array can be converted into n * m matrix: a[x] → matrix[x / m][x % m]
On the basis of this, binary search is applied to traverse throught the matrix and find the target element
 */

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length;
        int columns = matrix[0].length;
        int start = 0, end = rows*columns-1;

        while(start<=end)
        {
            int mid = start + (end-start)/2;
            int i = mid/columns;
            int j = mid%columns;

            if(matrix[i][j]==target)
            {
                return true;
            }
            else if(matrix[i][j]>target)
            {
                end = mid-1;
            }
            else 
            {
                start = mid+1;
            }
        }
        return false;
    }
}

File: solution1.py
# Time Complexity : O(log(mn)), where m and n are matrix dimensions
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach


def search_matrix(matrix, target) -> bool:

    m=len(matrix)
    # if matrix is empty
    if m==0:
        return False
    n=len(matrix[0])
    # running binary search on sorted matrix 
    left,right=0,m*n-1

    while left<=right:
        mid_idx=(left+right)//2
        # visualizing a sorted matrix as a sorted array where row and col are as follows
        mid=matrix[mid_idx//n][mid_idx%n]
        if target==mid:
            return True
        else:
            if target < mid:
                right=mid_idx-1
            else:
                left=mid_idx+1
    return False